extensions/warp-ipfs/src/store/files.rs

```diff
@@ -84,73 +74,420 @@ impl FileStore {
+        let _guard = Arc::new(token.clone().drop_guard());
 
-                    if signaled {
-                        Poll::Ready(Some(()))
-                    } else {
-                        Poll::Pending
-                    }
-                });
+        tokio::spawn(async move {
+            tokio::select! {
+                _ = task.run().instrument(span) => {}
+                _ = token.cancelled() => {}
+            }
+        });
+
+        Ok(FileStore {
+            index,
+            config,
+            path,
+            tx,
+            _guard,
+        })
+    }
+}
+
+impl FileStore {
+    pub fn modified(&self) -> DateTime<Utc> {
+        self.index.modified()
+    }
+
+    pub fn root_directory(&self) -> Directory {
+        self.index.clone()
+    }
+
+    /// Get the current directory that is mutable.
+    pub fn current_directory(&self) -> Result<Directory, Error> {
+        self.open_directory(&self.get_path().to_string_lossy())
+    }
+
+    /// Returns a mutable directory from the filesystem
+    pub fn open_directory(&self, path: &str) -> Result<Directory, Error> {
+        match path.trim().is_empty() {
+            true => Ok(self.root_directory()),
+            false => self
+                .root_directory()
+                .get_item_by_path(path)
+                .and_then(|item| item.get_directory()),
+        }
+    }
+
+    /// Current size of the file system
+    pub fn current_size(&self) -> usize {
+        self.root_directory().size()
+    }
+
+    pub fn max_size(&self) -> usize {
+        self.config.max_storage_size.unwrap_or(1024 * 1024 * 1024)
+    }
+
+    pub fn set_path(&mut self, path: PathBuf) {
+        *self.path.write() = path;
+    }
+
+    pub fn get_path(&self) -> PathBuf {
+        PathBuf::from(self.path.read().to_string_lossy().replace('\\', "/"))
+    }
+
+    pub async fn put(
+        &mut self,
+        name: impl Into<String>,
+        path: impl Into<String>,
+    ) -> Result<ConstellationProgressStream, Error> {
+        let (tx, rx) = oneshot::channel();
+        let _ = self
+            .tx
+            .clone()
+            .send(FileTaskCommand::Put {
+                name: name.into(),
+                path: path.into(),
+                response: tx,
+            })
+            .await;
+        rx.await.map_err(anyhow::Error::from)?
+    }
+
+    pub async fn get(
+        &self,
+        name: impl Into<String>,
+        path: impl Into<String>,
+    ) -> Result<ConstellationProgressStream, Error> {
+        let (tx, rx) = oneshot::channel();
+        let _ = self
+            .tx
+            .clone()
+            .send(FileTaskCommand::Get {
+                name: name.into(),
+                path: path.into(),
+                response: tx,
+            })
+            .await;
+        rx.await.map_err(anyhow::Error::from)?
+    }
+
+    pub async fn put_buffer(
+        &mut self,
+        name: impl Into<String>,
+        buffer: impl Into<Vec<u8>>,
```

Comment:
Do we really want to be generic here? We know we we always want these specific types that allocate so we can just take those types directly. Also, if the caller accidentally passes a `String` as an `&str`, there will be unnecessary allocation/copying.

A better generic here might be `impl Into<Box<str>>` and `impl Into<Box<[u8]>>`. That would mean we don't need to allocate/copy for sending it over the channel and in some cases, [the conversion wouldn't require allocation/copying even](https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-From%3CVec%3CT,+A%3E%3E-for-Box%3C%5BT%5D,+A%3E).